server/src/domain/person/person.service.ts

```diff
@@ -117,4 +117,53 @@ export class PersonService {
 
     return true;
   }
+
+  async mergePerson(authUser: AuthUserDto, primaryPersonId: string, dto: MergePersonDto) {
+    for (const id of dto.ids) {
+      const person = await this.repository.getById(authUser.id, id);
+      if (!person) {
+        throw new BadRequestException();
+      }
+    }
+
+    const primaryPerson = await this.repository.getById(authUser.id, primaryPersonId);
+    if (!primaryPerson) {
+      throw new BadRequestException();
+    }
+
+    const mergeIds = dto.ids;
+    for (const mergePersonId of mergeIds) {
+      // Find and remove duplicated entry in asset_faces table
+      const identicalAssetIds = await this.repository.deleteIdenticalAssets(primaryPersonId, mergePersonId);
+
+      // Remove record of duplicated entry in asset_faces table
+      // and Typesense database belong to merge person
+      if (identicalAssetIds.length > 0) {
+        for (const assetId of identicalAssetIds) {
+          await this.jobRepository.queue({
+            name: JobName.SEARCH_REMOVE_FACE,
+            data: { assetId, personId: mergePersonId },
+          });
+        }
+        this.logger.debug(`Deleted ${identicalAssetIds.length} duplicated assets`);
+      }
+
+      const mergePerson = await this.repository.getById(authUser.id, mergePersonId);
+      if (!mergePerson) {
+        throw new BadRequestException();
+      }
```

Comment:
There is a bit of redundant code looking up these merge ids. Maybe we can only do it once here, but instead of throwing an error return a list of results, this one being marked as failed, reason - not found. This is what I do for add/removing assets and I think it works pretty well. (domain/tag/tag.service.ts#addAssets and `AssetIdsResponseDto` is an example of what I am talking about). In general, we should have a discussion about "bulk requests" and how we want to handle them, but I think we should be consistent.